File: lib/asset/directory.py
from .url import URL


class Directory:
    """Directory object holds directory related data"""

    def __init__(self, level, url):
        """Initializes Directory object

        :param int level: depth level of directory
        :param URL url: URL object of directory location
        """
        self.depth_level = level
        self._url = url

    @property
    def url(self) -> str:
        return self._url.geturl()

    def pop_path(self) -> None:
        self._url.pop_path()

    def total_paths(self) -> int:
        total = self._url.path.count("/")
        return total

    def __eq__(self, other):
        return self.url == other.url

    def __ne__(self, other):
        return self.url != other.url

    def __hash__(self):
        return hash((self.url, self.depth_level))

File: app/src/main/java/xyz/zimtools/zyod/assets/Directory.java
package xyz.zimtools.zyod.assets;

import java.util.Objects;

/**
 * Holds information about a URL directory
 */
public final class Directory {
    private static final int FORWARD_SLASH_CHAR_CODE = 47;
    private int depthLevel;
    private final ODUrl url;

    public Directory(int level, ODUrl url) {
        this.depthLevel = level;
        this.url = url;
    }

    public int getDepthLevel() {
        return depthLevel;
    }

    /**
     * Sets depth level to zero.
     */
    public void resetDepthLevel() {
        this.depthLevel = 0;
    }

    public void popPath() {
        this.depthLevel -= 1;
        this.url.popPath();
    }

    /*
     * Total number of directories found in the current URL address.
     * */
    public long totalFolders() {
        return this.url.getPath().chars().filter(charNum -> charNum == FORWARD_SLASH_CHAR_CODE).count();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Directory directory = (Directory) o;
        return depthLevel == directory.depthLevel && url.equals(directory.url);
    }

    @Override
    public int hashCode() {
        return Objects.hash(depthLevel, url);
    }


    @Override
    public String toString() {
        return this.url != null ? this.url.toString() : "";
    }

}